src/test/scala/chiselTests/PrefixSpec.scala

```diff
@@ -0,0 +1,213 @@
+
+      val foo_inst = withModulePrefix("Foo") {
+        Instantiate(new AddOne(width))
+      }
+
+      val bar_inst = withModulePrefix("Bar") {
+        Instantiate(new AddOne(width))
+      }
+
+      // np: no prefix
+      val np_inst = Instantiate(new AddOne(width))
+
+      foo_inst.in := in
+      bar_inst.in := foo_inst.out
+      out := bar_inst.out
+      np_inst.in := in
+    }
+
+    val chirrtl = emitCHIRRTL(new Top)
+
+    val lines = """
+      module Foo_AddOne :
+      module Bar_AddOne :
+      module AddOne :
+      public module Top :
+        inst np_inst of AddOne
+        inst foo_inst of Foo_AddOne
+        inst bar_inst of Bar_AddOne
+        """.linesIterator.map(_.trim).toSeq
+
+    matchesAndOmits(chirrtl)(lines: _*)("AddOne_1")
+  }
+
+  it should "Instantiate should reference the same module definitions when instantiated with the same prefix" in {
+    class Top extends Module {
+      val width = 8
+      val in = IO(Input(UInt(width.W)))
+      val out = IO(Output(UInt(width.W)))
+      val foo_inst1 = withModulePrefix("Foo") {
+        Instantiate(new AddOne(width))
+      }
+
+      val foo_inst2 = withModulePrefix("Foo") {
+        Instantiate(new AddOne(width))
+      }
+
+      foo_inst1.in := in
+      foo_inst2.in := in
+      out := foo_inst1.out
+    }
+
+    val chirrtl = emitCHIRRTL(new Top)
+
+    val lines = """
+      module Foo_AddOne :
+      public module Top :
+        inst foo_inst1 of Foo_AddOne
+        inst foo_inst2 of Foo_AddOne
+        """.linesIterator.map(_.trim).toSeq
+
+    matchesAndOmits(chirrtl)(lines: _*)("AddOne_1", "Bar_AddOne")
+  }
+
+  it should "Memories work" in {
+    class Top extends Module {
+      val io = IO(new Bundle {
+        val enable = Input(Bool())
+        val write = Input(Bool())
+        val addr = Input(UInt(10.W))
+        val dataIn = Input(UInt(8.W))
+        val dataOut = Output(UInt(8.W))
+      })
+
+      val mem = withModulePrefix("Foo") {
+        SyncReadMem(1024, UInt(8.W))
+      }
+
+      mem.write(io.addr, io.dataIn)
+      io.dataOut := mem.read(io.addr, io.enable)
+    }
+
+//    val sv = ChiselStage.emitSystemVerilog(new Top)
+//    println(sv)
+
+    val chirrtl = emitCHIRRTL(new Top, args = Array("--full-stacktrace"))
+    println(chirrtl)
+
+    val lines = """
+      module mem_1024x8(
+      module Top(
+        """.linesIterator.map(_.trim).toSeq
+
+//    matchesAndOmits(sv)(lines: _*)()
+  }
+
+  it should "Definitions that appear within withModulePrefix get prefixed" in {
+    class Top extends Module {
+      val dfn = withModulePrefix("Foo") {
+        Definition(new AddOne(8))
+      }
+
+      val addone = Instance(dfn)
+    }
+
+    val chirrtl = emitCHIRRTL(new Top)
+    println(chirrtl)
+
+    val lines = """
+  module Foo_AddOne
+  module Top
+        """.linesIterator.map(_.trim).toSeq
+
+    matchesAndOmits(chirrtl)(lines: _*)()
+  }
+
+  it should "withModulePrefix does not affect ExtModules" in {
```

Comment:
```suggestion
  it should "withModulePrefix does not automatically affect ExtModules" in {
```

Reply:
should add another test that user can override desiredName with Module.prefix or whatever the user API is and then it *can* affect the ExtModule

Reply:
I agree we need this test because I think it will highlight a slight problem--`Module.currentModulePrefix` does not (I think) include a trailing separator, so how can you use it to properly generate `desiredName`?

We need tests for `Module.currentModulePrefix` and the `List` form too.